heft-plugins/heft-sass-plugin/src/SassProcessor.ts

```diff
@@ -92,243 +111,740 @@ export interface ISassConfiguration {
+        ...scssOptions,
+        url: record.url,
+        syntax: determineSyntaxFromFilePath(sourceFilePath)
       });
     } catch (err) {
       const typedError: Exception = err;
       const { span } = typedError;
 
-      // Extract location information and format into the error message until we have a concept
-      // of location-aware diagnostics in Heft.
-      throw new Error(`${typedError}(${span.start.column},${span.start.line}): ${typedError.message}`);
+      throw new FileError(`${typedError.sassMessage}\n${span.context ?? span.text}${typedError.sassStack}`, {
+        absolutePath: span.url ? heftUrlToPath(span.url.href ?? span.url) : 'unknown', // This property should always be present
+        line: span.start.line,
+        column: span.start.column,
+        projectFolder: this._options.buildFolder
+      });
     }
 
-    // Register any @import, @use files as dependencies.
+    // Register any @import files as dependencies.
+    record.dependencies.clear();
     for (const dependency of result.loadedUrls) {
-      this.registerDependency(filePath, fileURLToPath(dependency as URL));
+      const dependencyPath: string = heftUrlToPath(dependency.href);
+      const dependencyRecord: IFileRecord = this._getOrCreateRecord(dependencyPath);
+      record.dependencies.add(dependencyRecord);
+      dependencyRecord.consumers.add(record);
+    }
+
+    let css: string = result.css.toString();
+    const contentHash: string = getContentsHash(sourceFilePath, css);
+    if (record.cssVersion === contentHash) {
+      // The CSS has not changed, so don't reprocess this and downstream files.
+      return;
     }
 
-    return result.css.toString();
+    record.cssVersion = contentHash;
+    const { cssOutputFolders, dtsOutputFolders, srcFolder, exportAsDefault, postProcessCssAsync } =
+      this._options;
+
+    // Handle CSS modules
+    let moduleMap: JsonObject | undefined;
+    if (record.isModule) {
+      const postCssModules: postcss.Plugin = cssModules({
+        getJSON: (cssFileName: string, json: JsonObject) => {
+          // This callback will be invoked during the promise evaluation of the postcss process() function.
+          moduleMap = json;
+        },
+        // Avoid unnecessary name hashing.
+        generateScopedName: (name: string) => name
+      });
+
+      const postCssResult: postcss.Result = await postcss
+        .default([postCssModules])
+        .process(css, { from: sourceFilePath });
+      css = postCssResult.css;
+    }
+
+    if (postProcessCssAsync) {
+      css = await postProcessCssAsync(css);
+    }
+
+    const relativeFilePath: string = path.relative(srcFolder, sourceFilePath);
+
+    const dtsContent: string = this._createDTS(moduleMap);
+
+    const writeFileOptions: IFileSystemWriteFileOptions = {
+      ensureFolderExists: true
+    };
+
+    const promises: Promise<void>[] = dtsOutputFolders.map((dtsOutputFolder: string) => {
```

Comment:
Just a general question, should these promises be handled using `Async.forEachAsync`? Or is there no/minimal impact to spinning up all these write promises?

Reply:
Concurrency is managed by the caller of this function. For simplicity I'll just do these writes serially.